mrp_bom_hierarchy/model/mrp_bom.py

```diff
@@ -0,0 +1,261 @@
+                        ("product_tmpl_id", "=", bom.product_tmpl_id.id),
+                        ("id", "!=", bom.id),
+                    ],
+                )
+            if bom_ids:
+                bom.product_has_other_bom = True
+            else:
+                bom.product_has_other_bom = False
+
+    @api.depends("bom_line_ids.bom_id", "product_id", "product_tmpl_id")
+    def _compute_bom_ids(self):
+        for bom in self:
+            parent_bom_line_ids = self.env["mrp.bom.line"]._bom_line_find(
+                product_tmpl=bom.product_id.product_tmpl_id or bom.product_tmpl_id,
+                product=bom.product_id,
+            )
+            if parent_bom_line_ids:
+                bom.bom_ids = parent_bom_line_ids.bom_id
+                bom.has_parent = True
+            else:
+                bom.bom_ids = False
+                bom.has_parent = False
+
+    @api.depends("bom_line_ids.bom_id", "bom_line_ids.product_id")
+    def _compute_bom_lines(self):
+        for bom in self:
+            bom_line_ids = bom.bom_line_ids
+            bom.bom_lines = bom_line_ids.child_bom_id
+            bom.has_child = bool(bom.bom_lines)
+
+    def _search_has_child(self, operator, value):
+        if operator not in ["=", "!="]:
+            raise UserError(_("This operator is not supported"))
+        if value == "True":
+            value = True
+        elif value == "False":
+            value = False
+        if not isinstance(value, bool):
+            raise UserError(_("Value should be True or False (not %s)") % value)
+        ops = {"=": py_operator.eq, "!=": py_operator.ne}
+        ids = []
+        for bom in self.search([]):
+            if ops[operator](value, bom.has_child):
+                ids.append(bom.id)
+        return [("id", "in", ids)]
+
+    def _search_has_parent(self, operator, value):
+        if operator not in ["=", "!="]:
+            raise UserError(_("This operator is not supported"))
+        if value == "True":
+            value = True
+        elif value == "False":
+            value = False
+        if not isinstance(value, bool):
+            raise UserError(_("Value should be True or False (not %s)") % value)
+        ops = {"=": py_operator.eq, "!=": py_operator.ne}
+        ids = []
+        for bom in self.search([]):
+            if ops[operator](value, bom.has_parent):
+                ids.append(bom.id)
+        return [("id", "in", ids)]
+
+    @api.depends(
+        "product_id",
+        "product_id.default_code",
+        "product_id.product_tmpl_id.default_code",
+        "product_tmpl_id.default_code",
+    )
+    def _compute_internal_reference(self):
+        for bom in self:
+            bom.product_default_code = (
+                bom.product_id.default_code
+                or bom.product_id.product_tmpl_id.default_code
+                or bom.product_tmpl_id.default_code
+            )
+
+    bom_lines = fields.One2many("mrp.bom", compute="_compute_bom_lines")
```

Comment:
better name would be `child_bom_ids` IMO, current `bom_lines` name it is really misleading.